### ISSUE TITLE
Proposal: Nested Destructuring

### ISSUE THREAD

**Issue by Turbo87** [eslint/eslint#7973]
The purpose of this proposal is to introduce an option to warn about nested destructuring, which can be hard to read and understand. This is in line with the original proposal in https://github.com/eslint/eslint/issues/6053.

**What rule do you want to change?**

`prefer-destructuring`

**Does this change cause the rule to produce more or fewer warnings?**

It will produce more warnings.

**How will the change be implemented? (New option, new default behavior, etc.)?**

In the current state the rule can only be configured to emit warnings if a destructuring is possible and it doesn't care if the destructuring is nested. It is however not possible to do the opposite and warn if destructuring is used, and it is also not possible to warn only when nested destructuring is used.

I would like to propose to deprecate the `true/false` settings of `array/object` in the rule options and replace them with `always/flat/never`.

- `always` should warn when destructuring is possible
- `flat` should warn when nested destructuring is used, or flat destructuring is possible
- `never` should warn when destructuring is used

I would personally prefer it if the default was changed to `flat` for both `array` and `object`, but that would be a breaking change, so likely not a good idea. An alternativ however is introducing this as a new `destructuring` rule, and deprecate the `prefer-destructuring` rule completely.

**Please provide some example code that this change will affect:**

```js
   const { inject: { service }, Mixin, assert, isPresent } = Ember;

   let { a: [{foo: f}] } = obj
```

**What does the rule currently do for this code?**

Nothing, as the rule is currently not warning about destructuring.

**What will the rule do after it's changed?**

```js
// Do not use nested destructuring (at 1:17)
   const { inject: { service }, Mixin, assert, isPresent } = Ember;
// ----------------^

// Do not use nested destructuring (at 3:10)
   let { a: [{foo: f}] } = obj
// ---------^
```

---

<details>
 <summary><code>no-nested-destructuring</code> proposal</summary>

**Please describe what the rule should do:**

The rule should warn about the use of nested destructuring.

**What category of rule is this? (place an "X" next to just one item)**

[x] Enforces code style
[ ] Warns about a potential error
[x] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
// Do not use nested destructuring (at 1:17)
   const { inject: { service }, Mixin, assert, isPresent } = Ember;
// ----------------^

// Do not use nested destructuring (at 3:10)
   let { a: [{foo: f}] } = obj
// ---------^
```

First draft implementation is available at https://astexplorer.net/#/vTZ1hYHzaZ

**Why should this rule be included in ESLint (instead of a plugin)?**

Using nested destructuring can often lead to code that is harder to read and comprehend compared to basic variable assignments or using multiple non-nested destructurings in a row. I think this is a rule that should be included in ESLint itself to increase awareness of the problem and help everyone write easier to understand code. This could obviously be implemented in a plugin but ultimately I think this is something that the majority of JS/ES6 developers would use.
</details>

**Comment by platinumazure** [comment]
This should probably be done as an enhancement to `prefer-destructuring`, which currently favors the nested destructure in most cases.

**Comment by Turbo87** [comment]
@platinumazure I had thought about that option too. Might make more sense.

What's the further process here? Wait for other members to comment? Write a PR that modifies the `prefer-destructuring` rule?

a few more questions:

- should `prefer-destructuring` be deprecated for a `destructuring` rule that has `always`, `never`, `flat` options or something like that?
- if not, how should the options for `prefer-destructuring` be adjusted to configure the flat/non-nested style?

**Comment by vitorbal** [comment]
@Turbo87 the next step would be updating the issue title and description to use the "rule enhancement" template instead ([see here](https://github.com/eslint/eslint/blob/master/templates/rule-change-proposal.md)).

The proposal would need 3 +1's from ESLint team members and also one ESLint team member willing to champion it.

**Comment by Turbo87** [comment]
@platinumazure @vitorbal I've updated the proposal and also included a link to the original rule proposal (https://github.com/eslint/eslint/issues/6053) which also mentions that nested destructuring should probably be warned about by default.

/cc @zxqfox @kaicataldo @markelog I'm putting you on CC as you mentioned in the original proposal that you would also prefer flat destructuring to be the default.

If you need anything else from me, please let me know!

**Comment by vitorbal** [comment]
@Turbo87 thanks for the update! Here's my opinion: I am totally on board with an option to disallow nested destructuring, as I also believe they are hard to read and would like to forbid them.
I am not totally convinced by the option schema though, but I also can't come up with a better proposal at the moment.
Also, please note that instead of deprecating the true/false values, we should probably still support it in a backwards compatible way, to avoid any breaking changes.

**Comment by Turbo87** [comment]
> Also, please note that instead of deprecating the true/false values, we should probably still support it in a backwards compatible way, to avoid any breaking changes.

yes, that is what I meant by deprecating. 👍

**Comment by kaicataldo** [comment]
I think an option like this makes sense, though, thoughts on making the option a bit more generic/configurable by being able to set a hard limit on the number of levels of destructuring is allowed (setting it to 1 would be the same as "flat")? Seems like a worthwhile thing to explore.

**Comment by vitorbal** [comment]
@kaicataldo yep, I think having it configurable by number of destructuring levels makes sense (y).

**Comment by vitorbal** [comment]
@Turbo87 would you be willing to submit a PR for this change if it gets accepted?

**Comment by Turbo87** [comment]
@vitorbal sure, that's why I asked 😉

**Comment by joncursi** [comment]
> yep, I think having it configurable by number of destructuring levels makes sense

Agreed 100%. I want to use destructuring, but I also don't want my code to turn into a soup of curly braces. Having a configurable option for depth would give some much-needed flexibility.

**Comment by joncursi** [comment]
Or perhaps, a setting that allows you to specify when object destructuring should kick in, such as after `this`.

```
// OK
const something = process.env.variable;

// NOT OK
const something = this.props.something;
```

**Comment by platinumazure** [comment]
Vaguely related proposal around number of extracted properties: #8113 

Not directly related, just wanted to link together the different proposed approaches for solving the basic problem of "too much destructuring".

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get enough support from the team and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after a long time tend to never do it, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.

By the way, it should be possible to prevent nested destructuring by using a configuration of the [`no-restricted-syntax`](http://eslint.org/docs/rules/no-restricted-syntax) rule.
